File: apps/desktop/src/renderer/accountNumber.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { formatAccountNumberInput, normalizeAccountNumber } from "./accountNumber.ts";

test("legacy digit-only tokens pass through untouched", () => {
  assert.equal(normalizeAccountNumber("0123456789012345"), "0123456789012345");
  assert.equal(formatAccountNumberInput("0123456789012345"), "0123456789012345");
  assert.equal(formatAccountNumberInput("0123"), "0123");
});

test("normalize uppercases, strips separators and folds O/I/L", () => {
  assert.equal(
    normalizeAccountNumber("abcd-efgh jkmn-pqrs-tvwx-yz01"),
    "ABCDEFGHJKMNPQRSTVWXYZ01"
  );
  assert.equal(normalizeAccountNumber("oil"), "011");
  assert.equal(normalizeAccountNumber("  ABCD-EFGH  "), "ABCDEFGH");
});

test("normalize is idempotent", () => {
  const once = normalizeAccountNumber("abcd-efgh-jkmn-pqrs-tvwx-yz01");
  assert.equal(normalizeAccountNumber(once), once);
});

test("format groups in blocks of four with no trailing dash", () => {
  assert.equal(formatAccountNumberInput("abcdefghjkmnpqrstvwxyz01"), "ABCD-EFGH-JKMN-PQRS-TVWX-YZ01");
  assert.equal(formatAccountNumberInput("abcd"), "ABCD");
  assert.equal(formatAccountNumberInput("abcde"), "ABCD-E");
});

test("format is stable when re-applied to its own output", () => {
  const once = formatAccountNumberInput("abcdefghjkmnpqrstvwxyz01");
  assert.equal(formatAccountNumberInput(once), once);
});

test("format does not reject over-long or unexpected characters", () => {
  assert.equal(formatAccountNumberInput("abcdefghjkmnpqrstvwxyz0123456"), "ABCD-EFGH-JKMN-PQRS-TVWX-YZ01-2345-6");
  assert.equal(normalizeAccountNumber("ab!cd"), "AB!CD");
});

test("empty input stays empty", () => {
  assert.equal(normalizeAccountNumber(""), "");
  assert.equal(formatAccountNumberInput(""), "");
});

File: apps/android/core/src/test/kotlin/org/pangeavpn/core/AccountNumberTest.kt
package org.pangeavpn.core

import kotlin.test.Test
import kotlin.test.assertEquals
import org.pangeavpn.core.util.formatAccountNumberInput
import org.pangeavpn.core.util.normalizeAccountNumber

class AccountNumberTest {

    @Test
    fun `legacy digit-only tokens pass through untouched`() {
        assertEquals("0123456789012345", normalizeAccountNumber("0123456789012345"))
        assertEquals("0123456789012345", formatAccountNumberInput("0123456789012345"))
        assertEquals("0123", formatAccountNumberInput("0123"))
    }

    @Test
    fun `normalize uppercases, strips separators and folds O I L`() {
        assertEquals(
            "ABCDEFGHJKMNPQRSTVWXYZ01",
            normalizeAccountNumber("abcd-efgh jkmn-pqrs-tvwx-yz01"),
        )
        assertEquals("011", normalizeAccountNumber("oil"))
        assertEquals("ABCDEFGH", normalizeAccountNumber("  ABCD-EFGH  "))
    }

    @Test
    fun `normalize is idempotent`() {
        val once = normalizeAccountNumber("abcd-efgh-jkmn-pqrs-tvwx-yz01")
        assertEquals(once, normalizeAccountNumber(once))
    }

    @Test
    fun `format groups in blocks of four with no trailing dash`() {
        assertEquals("ABCD-EFGH-JKMN-PQRS-TVWX-YZ01", formatAccountNumberInput("abcdefghjkmnpqrstvwxyz01"))
        assertEquals("ABCD", formatAccountNumberInput("abcd"))
        assertEquals("ABCD-E", formatAccountNumberInput("abcde"))
    }

    @Test
    fun `format is stable when re-applied to its own output`() {
        val once = formatAccountNumberInput("abcdefghjkmnpqrstvwxyz01")
        assertEquals(once, formatAccountNumberInput(once))
    }

    @Test
    fun `nothing is rejected for length or charset`() {
        assertEquals("ABCD-EFGH-JKMN-PQRS-TVWX-YZ01-2345-6", formatAccountNumberInput("abcdefghjkmnpqrstvwxyz0123456"))
        assertEquals("AB!CD", normalizeAccountNumber("ab!cd"))
    }

    @Test
    fun `only ASCII digits count as a legacy token`() {
        // isDigit() would treat these as a legacy token and pass them through,
        // diverging from desktop's /^\d*$/ for the app's ar and fa locales.
        assertEquals("١٢٣٤", formatAccountNumberInput("١٢٣٤"))
        assertEquals("١٢٣٤-٥٦٧٨", formatAccountNumberInput("١٢٣٤٥٦٧٨"))
    }

    @Test
    fun `empty input stays empty`() {
        assertEquals("", normalizeAccountNumber(""))
        assertEquals("", formatAccountNumberInput(""))
    }
}
